ext/pdo/pdo_dbh.c

```diff
@@ -1055,18 +1063,25 @@ PHP_METHOD(PDO, query)
 	size_t statement_len;
 	zend_long fetch_mode;
 	zend_bool fetch_mode_is_null = 1;
-	zval *args = NULL;
-	uint32_t num_args = 0;
+	zval *arg1 = NULL;
+	HashTable *constructor_args = NULL;
 	pdo_dbh_object_t *dbh_obj = Z_PDO_OBJECT_P(ZEND_THIS);
 	pdo_dbh_t *dbh = dbh_obj->inner;
 
-	if (FAILURE == zend_parse_parameters(ZEND_NUM_ARGS(), "s|l!*", &statement, &statement_len, &fetch_mode, &fetch_mode_is_null, &args, &num_args)) {
+	if (FAILURE == zend_parse_parameters(ZEND_NUM_ARGS(), "s|l!z!h!", &statement, &statement_len,
```

Comment:
Please keep zpp variadic. IMHO this has made the implementation worse, not better, especially with the way you have to pass everything including argument numbers through to pdo_stmt_setup_fetch_mode now.

Reply:
The reason why I changed it is because I started by doing ``PDOStatement::fetchAll()`` which didn't use a ZPP variadic and so I based it around that. I will revert and change ``PDOStatement::fetchAll()`` to use variadics at the same time.

Reply:
If it's possible, I'd prefer to leave the fetchAll() refactoring until later, to make sure we can land this quickly.

Reply:
Right, will do so then